### DISCUSSION TITLE
SAP Luigi vs. SAPUI5 / OpenUI5

### DISCUSSION THREAD

**Discussion by pubmikeb** [SAP/luigi#2809] [Q&A] [answered]
Hello everyone,

SAP offers two frameworks for building front-end apps:
* [SAPUI5](https://ui5.sap.com/) / [OpenUI5](https://openui5.org/)
* [SAP Luigi](https://luigi-project.io)

Both look very similar and definitely (partly) share common codebase and design language.  

**My questions:**
 1. How one relates to another?
 2. What can be done with SAP Luigi, which is impossible with SAPUI5 / OpenUI5?
 3. In which scenarios one should be prefered over another?

**Comment by petermuessig** [comment, marked as answer]
Hi @pubmikeb ,

@hardl and I were just sitting together to collect some infos for you:

Roughly spoken: SAPUI5 / OpenUI5 is an Application UI Framework and SAP Luigi is a Micro Frontend Framework.

**Application UI Frameworks** provide an environment for developers to build applications and abstracts on the usage of the basic UI technologies (CSS Libraries, Web Components). Developers can focus on the application development and use existing UI elements and reuse components, but you explicitly decide on a programming model. Application UI Frameworks are enabling large scale application development with abstraction from HTML/CSS so that developers do not necessarily have web developer skills (although it is recommended to have some 😄 ). In case of SAPUI5 the development is embedded into the SAP ecosystem.

**Micro Frontend Frameworks** can be more seen as an application composition environment which stitches together individual micro frontends. It doesn't impose the UI framework to be used to develop one of these micro frontends but more provides an environment which establishes cross-micro-frontend communication channels, UI composition configurations, micro-frontend isolations, ... More details can be seen at https://martinfowler.com/articles/micro-frontends.html.

SAP Luigi composes micro frontends which are standalone web applications developed with any web UI framework such as SAPUI5 / OpenUI5 or something else. SAPUI5 applications or reuse components can be used within SAP Luigi to compose higher-level applications. Usually in a micro frontend architecture, higher-level applications will be split into different micro-frontends and developed by individual teams. Later the micro frontend framework (in our case SAP Luigi) connects the different micro frontends again.

Regarding 2) SAP Luigi is a different kind of framework to build and compose micro-frontend based applications out of different web applications built in different UI frameworks. SAPUI5 / OpenUI5 is also able to build different application parts in reuse components but there you have to stay in the SAPUI5 / OpenUI5 framework to compose the application. SAP Luigi also provides mechanism for isolation of the micro frontends which isn't the case for SAPUI5 / OpenUI5 application composition.

With SAPUI5 / OpenUI5 you can build web applications in all its depth. This isn't possible with Luigi as this is the micro frontend framework (the bracket) around the web applications. So, both are targeting two complete different scenarios and complement each other!

Regarding 3) SAP Luigi should be used for micro frontend compositions and SAPUI5 / OpenUI5 should be used for enterprise web application development. The question shouldn't be when should I use what, IMO the question should be how SAP Luigi complements SAPUI5 / OpenUI5 and how both can be used together.

Best regards,
Philipp (@hardl) and Peter

**Comment by pubmikeb** [comment]
@petermuessig and @hardl,  thanks for a such detailed answer.

I've also asked this question on SO — [SAP Luigi vs. SAPUI5 / OpenUI5](https://stackoverflow.com/questions/72942029/sap-luigi-vs-sapui5-openui5).
It would be great to repost your answer there as well.

**Reply by hardl** [reply]
@pubmikeb seems like the question on SO has been removed
